File: frontend/src/types/task.ts
import type { PageQuery } from './common'

export interface Task {
  id: number
  project_id: number
  project_name?: string
  parent_id?: number
  name: string
  owner_id: number
  owner_ids: number[]
  owner_names: string[]
  owner_name?: string
  project_manager_name?: string
  project_manager_id: number
  can_manage: boolean
  can_edit: boolean
  can_delete: boolean
  planned_start: string
  planned_end: string
  estimated_hours: number
  booked_hours: number
  status: string
  effective_status: string
  priority: string
  description?: string
  remark?: string
  children?: Task[]
}

export interface TaskPayload {
  project_id: number
  parent_id?: number
  name: string
  owner_ids: number[]
  planned_start: string
  planned_end: string
  estimated_hours: number
  description?: string
  remark?: string
}

export interface TaskQuery extends PageQuery {
  project_id?: number
  owner_id?: number
  status?: string
  department_id?: number
  organization_id?: number
  employee_no?: string
  name?: string
  managed_project_scope?: boolean
  personnel_keyword?: string
  organization_keyword?: string
}

File: backend/app/schemas/task.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import Field, model_validator

from app.schemas.common import ORMModel

TASK_STATUSES = {"not_started", "running", "completed"}
TASK_FILTER_STATUSES = TASK_STATUSES | {"delayed"}


class TaskBase(ORMModel):
    project_id: int
    parent_id: int | None = None
    name: str = Field(min_length=1, max_length=200)
    planned_start: date
    planned_end: date
    estimated_hours: Decimal = Field(default=Decimal("0.5"), gt=0, multiple_of=Decimal("0.5"))
    description: str | None = None
    remark: str | None = None

    @model_validator(mode="after")
    def validate_task(self):
        if self.planned_end < self.planned_start:
            raise ValueError("planned_end must be on or after planned_start")
        return self


class TaskCreate(TaskBase):
    owner_ids: list[int] = Field(min_length=1)


class TaskUpdate(ORMModel):
    parent_id: int | None = None
    name: str | None = Field(default=None, min_length=1, max_length=200)
    owner_ids: list[int] | None = Field(default=None, min_length=1)
    planned_start: date | None = None
    planned_end: date | None = None
    estimated_hours: Decimal | None = Field(default=None, gt=0, multiple_of=Decimal("0.5"))
    description: str | None = None
    remark: str | None = None


class TaskResponse(TaskBase):
    id: int
    owner_id: int
    owner_ids: list[int] = Field(default_factory=list)
    owner_names: list[str] = Field(default_factory=list)
    status: str
    priority: str = "medium"
    project_name: str | None = None
    owner_name: str | None = None
    project_manager_name: str | None = None
    project_manager_id: int | None = None
    can_manage: bool = False
    can_edit: bool = False
    can_delete: bool = False
    booked_hours: Decimal = Decimal("0")
    effective_status: str
    created_at: datetime
    updated_at: datetime
